AsyncEvaluator

emailweixu opened the pull request:
Fixes #1271 

The AsyncEvaluator creates a different process to evaluate the performance asynchronously so that it will not block the training. It also supports evaluation using parallel environments by setting TrainerConfig.num_eval_environments to number greater than one.

emailweixu committed:
AsyncEvaluator

hnyu commented on alf/metrics/metric.py:
```diff
@@ -58,14 +60,18 @@ def result(self):
+    def gen_summaries(self,
+                      train_step=None,
+                      step_metrics=(),
+                      other_steps: Dict[str, Number] = dict()):
```
Can this `other_steps` just be merged with `step_metrics`?

hnyu commented on alf/metrics/metric.py:
```diff
@@ -58,14 +60,18 @@ def result(self):
                 metric is generated against the global step.
             step_metrics: (Optional) Iterable of step metrics to generate summaries
                 against.
+            other_steps: A dictionary of steps to generate summaries against.
```
It's better to clarify that `other_steps` should directly contain integers instead of metrics, to differentiate from `step_metrics`.

emailweixu replied:
The type of other_steps indicates so.

hnyu commented on alf/trainers/evaluator.py:
```diff
@@ -0,0 +1,253 @@
+                if job.type == "eval":
+                    logging.info("Start evaluation")
+                    alf.summary.set_global_counter(job.global_counter)
+                    policy_trainer.Trainer._trainer_progress.update(
```
Why do we need to update trainer progress here?

emailweixu replied:
comment added.

hnyu commented on alf/trainers/evaluator.py:
```diff
@@ -0,0 +1,253 @@
+                    common.log_metrics(metrics)
+                    for metric in metrics:
+                        metric.gen_summaries(
+                            train_step=alf.summary.get_global_counter(),
```
Should this just be `job.global_counter`?

emailweixu replied:
They are same because the set_global_counter above.

hnyu commented on alf/trainers/evaluator.py:
```diff
@@ -0,0 +1,253 @@
+            type="eval",
+            env_steps=int(current_env_steps),
+            global_counter=int(alf.summary.get_global_counter()),
+            state_dict=algorithm.state_dict())
```
This limited number of EvalJob fields seems to make `gen_summaries` with other metrics not useful. Might have a field that store a list of steps?

emailweixu replied:
Changed env_steps to step_metric_values.

emailweixu committed:
Address comments

breakds commented:
Code change LGTM. The amortized GPU memory consumption should only slightly increase if my understanding is correct.

le-horizon commented on alf/trainers/evaluator.py:
```diff
@@ -0,0 +1,259 @@
+            global_counter=int(alf.summary.get_global_counter()),
+            state_dict=algorithm.state_dict())
+        self._job_queue.put(job)
+        self._done_queue.get()
```
When a previous eval job is still running and eval is called again, the worker is still busy working on the previous job and cannot yet take the job off the queue, so in this case, the evaluator will block on this done_queue.get()?

Maybe let's add a timer here to log to tensorboard how long this done_queue.get() is taking, so that we'll know that the eval process is blocking the training.

emailweixu replied:
added

le-horizon reviewed (commented):
Nice change, Wei!

emailweixu committed:
Also support synchronous evaluation.

In case that memory is a concern.
And write time (per training iteration) spent on evaluation to tensorboard.

breakds commented on alf/algorithms/config.py:
```diff
@@ -165,6 +166,8 @@ def __init__(self,
             num_eval_episodes (int) : number of episodes for one evaluation.
             num_eval_environments: the number of environments for evaluation.
+            async_eval: whether to do evaluation asynchronously in a different
+                process. Note that this more use more memory.
```
s/more/may/

emailweixu committed:
Fix typo and add more documentation.

hnyu commented on alf/trainers/evaluator.py:
```diff
@@ -38,29 +40,31 @@
+            self._done_queue = ctx.Queue()
+            self._worker = ctx.Process(
+                target=_worker,
+                args=(self._job_queue, self._done_queue, conf_file, num_envs,
```
Can we directly pass `config` to `_worker` instead of letting it parse `conf_file`?

emailweixu replied:
The subprocess is created using spawn method. Different from fork, config needs to be parsed again in the spawned subprocess anyway.
